src/main/scala/com/foursquare/common/thrift/bson/BranchingInputStream.scala

```diff
@@ -0,0 +1,21 @@
+// Copyright 2015 Foursquare Labs Inc. All Rights Reserved.
+
+package com.foursquare.common.thrift.bson
+
+import java.io.{ByteArrayInputStream, InputStream}
+
+/**
+ * version of ByteArrayInputStream that allows streams to be created
+ * from backing array without doing any copying
+ */
+class BranchingInputStream(
```

Comment:
"Branching" is a slightly confusing name, because the branch "rejoins" the main trunk after length bytes. Maybe "Forking"? Or if you don't want to change the names, I'm fine if you just clarify in the docstring.